File: app/src/main/java/com/adsamcik/signalcollector/utility/ChallengeDeserializer.java
package com.adsamcik.signalcollector.utility;

import com.adsamcik.signalcollector.data.Challenge;
import com.google.gson.JsonArray;
import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParseException;

import java.lang.reflect.Type;
import java.util.Random;

public class ChallengeDeserializer implements JsonDeserializer<Challenge> {
	@Override
	public Challenge deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
		JsonObject jobject = json.getAsJsonObject();
		JsonArray jDescVars = jobject.get("descVars").getAsJsonArray();
		String[] descVars = new String[jDescVars.size()];

		for (int i = 0; i < descVars.length; i++)
			descVars[i] = jDescVars.get(i).getAsString();

		JsonElement progressElement  = jobject.get("progress");
		float progress;
		if(progressElement == null)
			progress = 0;
		else
			progress = progressElement.getAsFloat();

		return new Challenge(
				Challenge.ChallengeType.valueOf(jobject.get("type").getAsString()),
				jobject.get("title").getAsString(),
				descVars,
				progress,
				jobject.get("difficulty").getAsInt());
	}
}

File: app/src/main/java/com/adsamcik/signalcollector/utility/ChallengeDeserializer.kt
package com.adsamcik.signalcollector.utility

import com.adsamcik.signalcollector.data.Challenge
import com.squareup.moshi.FromJson
import com.squareup.moshi.JsonReader
import com.squareup.moshi.JsonWriter
import com.squareup.moshi.ToJson
import java.io.IOException

/**
 * JsonDeserializer class that is needed to properly deserialize Challenge objects
 */
class ChallengeDeserializer {
    @FromJson fun fromJson(reader: JsonReader): Challenge? {
        var challengeType: Challenge.ChallengeType? = null
        var title: String? = null
        var descVars: Array<String>? = null
        var progress: Float? = null
        var difficulty: Int? = null

        reader.isLenient = true

        reader.beginObject()
        while (reader.hasNext()) {
            val name = reader.nextName()
            when (name) {
                "descVars" -> descVars = readStringArray(reader).toTypedArray()
                "progress" -> progress = reader.nextDouble().toFloat()
                "type" -> challengeType = Challenge.ChallengeType.valueOf(reader.nextString())
                "difficulty" -> difficulty = reader.nextInt()
                "title" -> title = reader.nextString()
            }
        }
        reader.endObject()

        if (challengeType == null || title == null || descVars == null || progress == null || difficulty == null)
            return null

        return Challenge(
                challengeType,
                title,
                descVars,
                progress,
                difficulty)
    }

    @Throws(IOException::class)
    private fun readStringArray(reader: JsonReader): List<String> {
        val strings = ArrayList<String>()

        reader.beginArray()
        while (reader.hasNext())
            strings.add(reader.nextString())

        reader.endArray()
        return strings
    }

    @ToJson fun toJson(writer: JsonWriter?, value: Challenge?) {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }
}
